cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+//
+// It is recommended, but not required, that provider-specific controllers add
+// finalizers to Machine objects so that they can be triggered on deletion to
+// release the necessary external resources, reporting any errors encountered.
+type Machine struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   MachineSpec
+	Status MachineStatus
+}
+
+type MachineSpec struct {
+	// This ObjectMeta will autopopulate the Node created. Use this to
+	// indicate what labels, annotations, name prefix, etc., should be used
+	// when creating the Node.
+	// +optional
+	metav1.ObjectMeta
+
+	// Provider-specific serialized configuration to use during node
+	// creation. It is recommended that providers maintain their own
+	// versioned API types that should be serialized/deserialized from this
+	// field, akin to component config.
+	// +optional
+	ProviderConfig string
+
+	// A list consisting of "Master" and/or "Node".
+	//
+	//                 +-----------------------+------------------------+
+	//                 | Master present        | Master absent          |
+	// +---------------+-----------------------+------------------------|
+	// | Node present: | Install control plane | Join the cluster as    |
+	// |               | and be schedulable    | just a node            |
+	// |---------------+-----------------------+------------------------|
+	// | Node absent:  | Install control plane | Invalid configuration  |
+	// |               | and be unscheduleable |                        |
+	// +---------------+-----------------------+------------------------+
+	Roles []string
+
+	// Versions of key software to use.
+	// +optional
+	Versions MachineVersionInfo
+
+	// To populate in the associated Node for dynamic kubelet config. This
+	// field already exists in Node, so any updates to it in the Machine
+	// spec will be automatially copied to the linked NodeRef from the
+	// status. The rest of dynamic kubelet config support should then work
+	// as-is.
+	// +optional
+	ConfigSource *corev1.NodeConfigSource
+}
+
+type MachineStatus struct {
+	// If the corresponding Node exists, this will point to its object.
+	// +optional
+	NodeRef *api.ObjectReference
+
+	// When was this status last observed
+	// +optional
+	LastUpdated metav1.Time
+
+	// Indicates whether or not the Machine is fully reconciled. When a
+	// controller observes that the spec has changed and no longer matches
+	// reality, it should update Ready to false before reconciling the
+	// state, and then set back to true when the state matches the spec.
+	Ready bool
+
+	// If set, indicates that there is a problem reconciling state, and
+	// will be set to a token value suitable for machine interpretation.
+	// +optional
+	ErrorReason *MachineStatusError
+
+	// +optional
+	// If set, indicates that there is a problem reconciling state, and
+	// will be set to a human readable string to indicate the problem.
+	ErrorMessage *string
+}
```

Comment:
Where does the Provider puts the status for the cloud resources it creates? `ProviderA` might create / update / delete `security_groups`, `keys` or `anything_related` for every machine it reconciles and those cloud resources are going to be completely different from `ProviderB`'s resources.

Reply:
Would `ErrorReason *MachineStatusError` and `ErrorMessage *string` make more sense in a list of structs?  What does the event data structure look like?  This almost appears to be a list of Error Events.

Reply:
```go
+
 +	// +optional
 +	// If set, indicates that there is a problem reconciling state, and
 +	// will be set to a human readable string to indicate the problem.
 +	ErrorMessage *string
```

Maybe this is a better question, since the struct may be below.  What is the `ErrorMessage`?  How does it relate to `ErrorReason`?

Reply:
We discussed this in one of the Cluster API breakout sessions, but I forgot to follow-up here: it's completely up to the controller as an implementation detail. A decent pattern is for the controller to add custom annotations to the Machines it's reconciling, to keep track of information about external resources it has created. The controller could also create its own ConfigMaps or CustomResources to have better control (and RBAC scoping) of its state, or even store state outside of the cluster if it makes sense.

Reply:
As for ErrorMessage / ErrorReason, this is definitely one of the warts of the design so far that I'd like to flesh out with any better approaches.

We did get the advice from Brian Grant and Eric Tune that the existing pattern of Conditions in object statuses is near-deprecated. As part of trying to bring several controllers to GA, a survey was sent out to understand how people use Conditions and how effective they are for the intent. The overall response was that having Conditions be a list of state transitions generally made them not useful for the kinds of checks people wanted to make against the Status, which are to answer the question "is this done yet?". This resulted in clients always just looking at the most recent Condition in the list and treating it as the current state, which on top of making the client logic more difficult, also made them deteriorate into Phases anyway (which are thoroughly deprecated).

So, they suggested two different patterns to replace Conditions:

1) If you really want a timeseries stream of state transitions, we should use Events.
2) For Status fields that we think clients will care to watch, we should just have fine-grained top-level entries (rather than lists) for the current state.

We're on the bleeding edge, so there aren't other parts of Kubernetes that have migrated off of Conditions yet, and we might be setting the precedents, or we may just need to slightly alter our types once better recommendations are in place.

The Error* fields were my attempt at (2), with the general guidelines that if a client modifies a field in the Machine's Spec, it should specifically watch for the corresponding field of the Machine Status see whether or not it has been reconciled, while watching the Error* fields for any errors that occur in the meantime. If you're updating the version of `kubelet` in the Spec, you should watch the corresponding field in the Status to know when it's been reconciled. This works decently with the Error* fields so long as you have a single controller responsible for the entirety of the Machine Spec, but breaks down somewhat if you want different controllers to handle different fields of the same object, or handle reconciling the same Machine under different circumstances.

For instance, one controller may be responsible whenever a full VM replacement is needed, while another may specialize in being able to update a VM in-place for certain Spec changes. It's not fantastic if they're unable to report errors separately, and instead have to overwrite the same fields in the Status. One mitigation is to always publish an Event on the Machine as well, so anyone who cares can still see the full stream of all errors. Another mitigation is to provide very strong guidance over what constitutes an error worth reporting in the Status.

I'll clarify this in the documentation, but I think it's a good idea for Machine.Status.Error* to be reserved for errors that are considered terminal, rather than transient. A terminal error would be something like the fact that the Machine.Spec has an invalid configuration, so the controller won't be able to make progress until someone fixes some aspect of it. Another terminal error would be if the machine-controller is getting Unauthorized or Permission Denied responses from the cloud provider it's calling to create/delete VMs -- it's likely going to require some manual intervention to fix IAM permissions or service credentials before it's able to do anything useful. However, any transient service failures can just be logged in the controller's output and/or added as Events to the Machine, since they should only represent delays in reconciliation and not errors that require intervention.

If two different controllers want to report terminal errors on the same Machine object, then I think it's okay that they are overwriting each other's errors in the Machine.Status, since they are both valid errors that need to be taken care of. By definition, neither controller is able to make progress until someone steps in and modifies the Machine.Spec or some aspect of the environment to fix the error, so someone will need to address both errors anyway (which may have the same underlying root cause). Based on the timing, they'll either see one error or the other first, and will have to fix it. Then, that error will disappear as the first controller gets unwedged, but a second error might replace it from the second controller, and the admin can take action on that as well.

We can figure out some way to represent both errors in the Status at the same time, but I'm not sure how much value there is in that over the above model, or above the cluster admin looking at the Machine events anyway.